File: p637e/average_of_levels.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import List
from local_packages.binary_tree import TreeNode


class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        ans, queue = [], [root] if root else []
        while queue:
            ans.append(sum([node.val for node in queue]) / len(queue))
            queue = [
                child for node in queue for child in (node.left, node.right) if child
            ]
        return ans


# TESTS
for tree, expected in [
    ("1", [1]),
    ("1,2,#,#,3,#,#", [1, 2.5]),
    ("1,2,3,4,#,#,#,#,#", [1, 2, 3, 4]),
    ("1,#,2,#,3,#,4,#,#", [1, 2, 3, 4]),
    ("1,#,2,3,4,#,#,#,#", [1, 2, 3, 4]),
    ("1,#,2,3,#,#,4,#,#", [1, 2, 3.5]),
    ("1,2,4,#,#,5,#,7,#,#,3,6,#,8,#,#,#", [1, 2.5, 5, 7.5]),
    ("3,9,#,#,20,15,#,#,7,#,#", [3, 14.5, 11]),
]:
    sol = Solution()
    actual = sol.averageOfLevels(TreeNode.deserialize(tree))
    print("Average of levels of", tree, "->", actual)
    assert actual == expected

File: p637e/average_of_levels.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevelsDFS = function (root) {
  let levels = [];
  const recPreorder = function (node, level) {
    if (node == null) return;
    if (levels[level]) levels[level].push(node.val);
    else levels.push([node.val]);
    recPreorder(node.left, level + 1);
    recPreorder(node.right, level + 1);
  };
  recPreorder(root, 0);
  return levels.map((level) => {
    let sum = level.reduce((prev, n) => prev + n, 0);
    return sum / level.length;
  });
};

// TEST
const f = (tree, expected) => [tree, expected];
[
  f('1', [1]),
  f('1,2,#,#,3,#,#', [1, 2.5]),
  f('1,2,3,4,#,#,#,#,#', [1, 2, 3, 4]),
  f('1,#,2,#,3,#,4,#,#', [1, 2, 3, 4]),
  f('1,#,2,3,4,#,#,#,#', [1, 2, 3, 4]),
  f('1,#,2,3,#,#,4,#,#', [1, 2, 3.5]),
  f('1,2,4,#,#,5,#,7,#,#,3,6,#,8,#,#,#', [1, 2.5, 5, 7.5]),
  f('3,9,#,#,20,15,#,#,7,#,#', [3, 14.5, 11]),
].forEach(([tree, expected]) => {
  const actual = averageOfLevelsDFS(Tree.deserialize(tree));
  console.log('Average of levels of', tree, '->', actual);
  console.assert(actual.toString() === expected.toString());
});
